File: Python/BOJ/2110 공유기 설치.py
import sys


def search(x):  # Parametric Search
    start = 1
    end = max(x)-min(x)  # 최대와 최소의 차
    result = 0
    while start <= end:
        middle = (start + end) // 2
        num = x[0]  # 최소로 지정
        count = 1

        for i in range(1, N):
            if x[i] >= num + middle:  # 최소 + middle보다 크면
                count += 1  # 공유기 갯수 증가
                num = x[i]

        if count >= C:
            start = middle + 1
            result = middle
        else:
            end = middle - 1
    return (result)


N, C = map(int, sys.stdin.readline().split())
x = [0]*N
for i in range(N):
    x[i] = int(sys.stdin.readline())
x.sort()
print(search(x))

File: Java/BOJ/2110 공유기 설치.java
import java.util.*;
import java.io.*;

public class Main {
	static int N;
	static int C;
	static int[] x;
	
	static int search() {
		int start = 1 ;
		int end = x[N-1] - x[0];
		int result = 0;
		while (start <= end) {
			int mid = (start+end)/2;
			int num = x[0];
			int count = 1;
			
			for (int i = 0 ; i < N; i ++) {
				if (x[i] >=  num + mid) {
					count++;
					num = x[i];
				}
			}
			if (count >= C) {
				start = mid + 1;
				result = mid;
			}
			else {
				end = mid- 1;
			}
		}
		return result;
	}
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        N = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());
        x = new int[N];
        for(int i = 0; i < N; i ++) {
        	x[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(x);
        System.out.println(search());
    }
}
